golang/internal/helper/image/image.go

```diff
@@ -0,0 +1,485 @@
+package image
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"runtime"
+	"time"
+
+	"github.com/dyrector-io/dyrectorio/golang/internal/logdefer"
+	"github.com/dyrector-io/dyrectorio/protobuf/go/agent"
+
+	"github.com/docker/cli/cli/command"
+	configtypes "github.com/docker/cli/cli/config/types"
+	"github.com/docker/cli/cli/trust"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	registrytypes "github.com/docker/docker/api/types/registry"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/docker/docker/registry"
+	"github.com/opencontainers/go-digest"
+	"github.com/rs/zerolog/log"
+	notaryClient "github.com/theupdateframework/notary/client"
+	"github.com/theupdateframework/notary/tuf/data"
+)
+
+type target struct {
+	name   string
+	digest digest.Digest
+	size   int64
+}
+
+// PullResponse is not explicit
+type PullResponse struct {
+	ID             string `json:"id"`
+	Status         string `json:"status"`
+	ProgressDetail struct {
+		Current int64 `json:"current"`
+		Total   int64 `json:"total"`
+	} `json:"progressDetail"`
+	Progress string `json:"progress"`
+}
+
+func GetRegistryURL(reg *string, registryAuth *RegistryAuth) string {
+	if registryAuth != nil {
+		return registryAuth.URL
+	} else if reg != nil {
+		return *reg
+	} else {
+		return ""
+	}
+}
+
+func GetRegistryURLProto(reg *string, registryAuth *agent.RegistryAuth) string {
+	if registryAuth != nil {
+		return registryAuth.Url
+	} else if reg != nil {
+		return *reg
+	} else {
+		return ""
+	}
+}
+
+func GetImageByReference(ctx context.Context, ref string) (*types.ImageSummary, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, err
+	}
+
+	images, err := cli.ImageList(ctx, types.ImageListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{Key: "reference", Value: ref}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(images) == 1 {
+		return &images[0], nil
+	}
+
+	return nil, errors.New("found more than 1 image with the same reference")
+}
+
+func Exists(ctx context.Context, logger io.StringWriter, expandedImageName string) (bool, error) {
```

Comment:
We should add something to the name to indicate that it represents a unique instance. What do you think? eg.: `ExistsOnlyOne` 

(just minor brainstorming)

Reply:
Image filter by default is in fact exact matching, therefor multi-matching is out of question.
It would be nice to have this as a comment above the function, what do you think?

Reply:
Okey for me.